### PR TITLE
Add dialogContext memory scope

### LINKED ISSUE #3210: Scope variable to know the current active dialog in the dialog stack
**Is your feature request related to a problem? Please describe.**
Since Bot Composer can easily become complex with many dialogs and flows, it would be very useful (at debug time) to know what the current active dialog is.
We try many combinations like:
dialog.name
this.dialog.name
this.dialogs[0].name
[...]
But none worked. As we have not been able to find the documentation on the properties of each scope variable, trying to find the solution by test and error can be crazy.

**Describe the solution you'd like**
Our workarround is to have a custom user.dialogPath variable that we set on each dialog first step with the current dialog name.
It would be nice to know if there is a var scope in bot composer that indicates the name of the current active dialog and, if it does not exist, create a new one.

**Describe alternatives you've considered**
Having a visual dialog stack for debug it could be amazing too.
If we have something like:
dialog.currentName = “IdentifyUserDialog”
dialog.currentStack = [ “IdentifyUserDialog”, “AssistanceDialog”, “MainDialog” ]
It means:
Current active dialog is “IdentifyUserDialog”
that was called from “AssistanceDialog”, that was called from “MainDialog”.

### LINKED ISSUE #3363: DialogStateManager.GetValue should have a non-generic overload that returns a string
**Is your feature request related to a problem? Please describe.**
Much of my stored conv/turn/dialog state is in strings. However, the methods available are:
```
- GetIntValue()
- GetBoolValue()
- GetValue<T>()
```

I've got a mid-sized bot and estimate I've typed `GetValue<string>()` about 50 times.

**Describe the solution you'd like**

I'd like either:
```
- string GetValue()   <- *best in show*
- string GetStringValue()
```

[enhancement]

### PR DESCRIPTION
**Pull request by tomlm**
Fixes #3210 
Fixes #3363 

Add new **DialogContext** memory scope with ids of 
* activeDialog
* parent
* stack

This allows you to determine if you are executing the context of a known dialog via expression
```=contains(dialogcontext.stack,'somedialogid')```
Also added 3 utility expression functions
* **isDialogActive('dialogId')** - returns true if a given dialog id is in the stack
* **isActionActive('actionId')** - simple alias for isDialogActive to match IA for actions.
* **hasPendingActions()** - returns true if there is nothing in the execution plan for the active dialog.

### CODE REVIEW

**Comment by vishwacsena** [comment]
@tomlm I'm not seeing Ids of input actions that are still pending.. e.g. I do not see 'askForName' in the stack for this dialog when the user says 'why' when the bot is asking for user name. 

```C#
var testDialog = new AdaptiveDialog("rootDialog")
{
    AutoEndDialog = false,
    Generator = new TemplateEngineLanguageGenerator(),
    Recognizer = new RegexRecognizer()
    {
        Intents = new List<IntentPattern>()
        {
            new IntentPattern()
            {
                Intent = "why",
                Pattern = "why"
            }
        }
    },
    Triggers = new List<OnCondition>()
    {
        new OnBeginDialog()
        {
            Actions = new List<Dialog>()
            {
                new TextInput()
                {
                    Id = "askForName",
                    Prompt = new ActivityTemplate("What is your name?"),
                    Property = "user.name"
                },
                new SendActivity("I have ${user.name}")
            }
        },
        new OnIntent()
        {
            Intent = "why",
            Actions = new List<Dialog>()
            {
                new SendActivity()
                {
                    Id = "Self",
                    Activity = new ActivityTemplate("I have ${join(dialogContext.stack, ' -> ')}")
                },
                new IfCondition()
                {
                    Condition = "dialogContext.activeDialog == 'askForName'",
                    Actions = new List<Dialog>()
                    {
                        new SendActivity("I need your name to complete the sample")
                    },
                    ElseActions = new List<Dialog>()
                    {
                        new SendActivity("I just need the info..")
                    }
                }
            }
        }
    }
};
```

**Comment by tomlm** [comment]
The issue is that you can't evaluate the context AFTER you have already interrupted.  You need to do it on the intent as a condition....to decide to interrupt... like this:
```C#
new OnIntent()
{
  Intent = "why",
  Condition = "contains(dialogcontext.stack, 'askForName')",
  Actions = new List<Dialog>()
  {
    new SendActivity("I need your name to complete the sample")
  }
},
new OnIntent()
{
  Intent = "why",
  Actions = new List<Dialog>()
  {
    new SendActivity("why what?")
  }
}
```

**Comment by vishwacsena** [comment]
@tomlm I spent some more time on this today and found the following - 
1. [Issue/ bug] EditActions --> AppendActions does not work. See OnIntent() for joke in the "root" dialog below. Basically if you try profile -> joke -> vishwac -> 37, the expectation is that the bot will come back with the joke after the userProfileDialog is done but it does not. 
2. [Odd] I wanted to have a trigger such that it will only work if we are in `turn.0` (no child dialog is in the stack). The expression I had to write to achieve that was `count(dialogContext.state) == 2` which is very wired. Is there a way we can add a prebuilt function that returns boolean true if there is nothing else in the stack? The reason I had to do to two was in my example, I had the rootDialog sit in a component dialog which I guess counts as 2 dialogs. e.g. it would be cleaner to write`Condition = "noDialogsActive()"`
3. [Odd] Having a set of actions be re-used by multiple conditions could get better with a helperFunction. I had to write something like this `contains(dialogContext.stack, 'askForName') || contains(dialogContext.stack, 'askForAge')` and this would cause much repetition. Could we instead do e.g. `Condition = "IsDialogActive('askForName') || IsDialogActive('askForAge')"`

Here's the full sample I wrote .. 

```C# 
var rootDialog = new AdaptiveDialog("root")
{
    AutoEndDialog = false,
    Generator = new TemplateEngineLanguageGenerator(),
    Recognizer = new RegexRecognizer()
    {
        Intents = new List<IntentPattern>()
        {
            new IntentPattern()
            {
                Intent = "profile",
                Pattern = "profile"
            },
            new IntentPattern()
            {
                Intent = "joke",
                Pattern = "joke"
            },
            new IntentPattern()
            {
                Intent = "start",
                Pattern = "start"
            }
        }
    },
    Triggers = new List<OnCondition>()
    {
        new OnBeginDialog()
        {
            Actions = new List<Dialog>()
            {
                new SendActivity("Say profile | joke | start to get started")
            }
        },

        // joke is always available if it is a turn.0 trigger. Joke is not available otherwise.
        new OnIntent()
        {
            Intent = "joke",
            Condition = "count(dialogContext.state) == 2",
            Actions = new List<Dialog>()
            {
                new BeginDialog()
                {
                    Dialog = "jokeDialog"
                }
            }
        },

        // queue up joke if we are in a sensitive multi-turn flow
        new OnIntent()
        {
            Intent = "joke",
            Condition = "contains(dialogContext.stack, 'askForName') || contains(dialogContext.stack, 'askForAge')",
            Actions = new List<Dialog>()
            {
                new EditActions()
                {
                    ChangeType = ActionChangeType.InsertActions,
                    Actions = new List<Dialog>()
                    {
                        new SendActivity("Sure. I will get you a joke after I get your name..")
                    }
                },
                new EditActions()
                {
                    ChangeType = ActionChangeType.AppendActions,
                    Actions = new List<Dialog>()
                    {
                        new SendActivity("Here is the joke you asked for.."),
                        new BeginDialog()
                        {
                            Dialog = "jokeDialog"
                        }
                    }
                }
            }
        },

        // Only possible if in turn.0. Not possible if this is an interruption
        new OnIntent()
        {
            Intent = "start",
            Condition = "count(dialogContext.stack) == 2",
            Actions = new List<Dialog>()
            {
                new SendActivity("In start .. this is not possible as an interruption...")
            }
        },

        // Do not do this if childDialog is already in the stack => child cannot self interrupt itself
        new OnIntent()
        {
            Intent = "profile",
            Condition = "!contains(dialogContext.stack, 'userProfileDialog')",
            Actions = new List<Dialog>()
            {
                new SendActivity("In profile .. this is always possible and I handle it immediately."),
                new BeginDialog()
                {
                    Dialog = "userProfileDialog"
                }
            }
        }
    }
};

var jokeDialog = new AdaptiveDialog("jokeDialog")
{
    Generator = new TemplateEngineLanguageGenerator(),
    Triggers = new List<OnCondition>()
    {
        new OnBeginDialog()
        {
            Actions = new List<Dialog>()
            {
                new SendActivity("Sure.. Here's the joke .. ha ha ha :) ")
            }
        }
    }
};

var userProfileDialog = new AdaptiveDialog("userProfileDialog")
{
    AutoEndDialog = false,
    Generator = new TemplateEngineLanguageGenerator(),
    Recognizer = new RegexRecognizer()
    {
        Intents = new List<IntentPattern>()
        {
            new IntentPattern()
            {
                Intent = "why",
                Pattern = "why"
            },
            new IntentPattern()
            {
                Intent = "no",
                Pattern = "no"
            }
        }
    },
    Triggers = new List<OnCondition>()
    {
        new OnBeginDialog()
        {
            Actions = new List<Dialog>()
            {
                new SendActivity("In profile dialog..."),
                new TextInput()
                {
                    Id = "askForName",
                    Prompt = new ActivityTemplate("What is your name?"),
                    Property = "user.name"
                },
                new SendActivity("I have ${user.name}"),
                new TextInput()
                {
                    Id = "askForAge",
                    Prompt = new ActivityTemplate("What is your age?"),
                    Property = "user.age"
                },
                new SendActivity("I have ${user.age}")
            }
        },
        new OnIntent()
        {
            Intent = "why",
            Condition = "contains(dialogContext.stack, 'askForName')",
            Actions = new List<Dialog>()
            {
                new SendActivity("I need your name to address you correctly"),
            }
        },
        new OnIntent()
        {
            Intent = "why",
            Condition = "contains(dialogContext.stack, 'askForAge')",
            Actions = new List<Dialog>()
            {
                new SendActivity("I need your age to provide relevant product recommendations")
            }
        },
        new OnIntent()
        {
            Intent = "why",
            Actions = new List<Dialog>()
            {
                new SendActivity("I need your information to complete the sample..")
            }
        },
        new OnIntent()
        {
            Intent = "no",
            Actions = new List<Dialog>()
            {
                new SetProperties()
                {
                    Assignments = new List<PropertyAssignment>()
                    {
                        new PropertyAssignment()
                        {
                            Property = "user.name",
                            Value = "Human"
                        },
                        new PropertyAssignment()
                        {
                            Property = "user.age",
                            Value = "30"
                        }
                    }
                }
            }
        },
        new OnIntent()
        {
            Intent = "no",
            Condition = "contains(dialogContext.stack, 'askForName')",
            Actions = new List<Dialog>()
            {
                new SetProperty()
                {
                    Property = "user.name",
                    Value = "Human"
                }
            }
        },
        new OnIntent()
        {
            Intent = "no",
            Condition = "contains(dialogContext.stack, 'askForAge')",
            Actions = new List<Dialog>()
            {
                new SetProperty()
                {
                    Property = "user.age",
                    Value = "30"
                }
            }
        }
    }
};
```

**Comment by vishwacsena** [comment]
@tomlm the isDialogActive() as well as hasPendingActions() functions work great. I have one issue and one suggestion. 

1. The EditAction with append change type still does not work for me. I see that you have a test but can you try to add a test that goes across dialogs? E.g. Dialog A starts dialog B and on interruption/ consultation dialog A does an editAction to append dialog C? This is where I see that the plan does not execute dialog C after dialog B is done. 
2. Given we have dialog and actions as two separate concepts, can we add an alias `isActionActive()` to `isDialogActive()`? It feels odd to write isDialogActive(ActionId) although actions are dialogs, that is not the experience/ mental model/ IA.

Rest look great! Thanks!

**Comment by tomlm** [comment]
The issue with your sample is the that **userProfileDialog.AutoEndDialog = false**, which means that it doesn't end when you complete, and so never executes the action appended to the plan. 
Setting that to true causes your example to work.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Functions/IsDialogActiveFunction.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        public const string Alias = "isActionActive";
+
+        public IsDialogActiveFunction()
+            : base(Name, Function, ReturnType.Boolean)
+        {
+        }
+
+        private static (object value, string error) Function(Expression expression, IMemory state, Options options)
+        {
+            if (state.TryGetValue("dialogcontext.stack", out object stackArray))
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Functions/IsDialogActiveFunction.cs:35]: Should this be 

```suggestion
            if (state.TryGetValue($"{ScopePath.DialogContext}.{DialogContextMemoryScope.Stack}", out object stackArray))
```

I worry about these being hard to change or track. Defining the constants might help? Thoughts? (note that I assumed the existence of a constant for "stack" also.

File: libraries/Microsoft.Bot.Builder.Dialogs/Memory/Scopes/DialogContextMemoryScope.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                    {
+                        stack.Add(item.Id);
+                    }
+                }
+
+                currentDc = currentDc.Parent;
+            }
+
+            // top of stack is stack[0]. 
+            memory["stack"] = stack;
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Memory/Scopes/DialogContextMemoryScope.cs:60]: Do these 3 strings deserve constants? I see stack referenced above.

**Pull request merged**